java/feathub-connectors/flink-connectors/flink-connector-prometheus/src/main/java/com/alibaba/feathub/flink/connectors/prometheus/MapMetricUpdater.java

```diff
@@ -43,24 +45,38 @@ public MapMetricUpdater(String name, Map<String, String> labels, CollectorRegist
                         .register(registry);
     }
 
+    private String[] getLabelValues(String mapKey) {
+        final ArrayList<String> labelValues = new ArrayList<>(this.labels.size());
+        for (String labelValue : labels.values()) {
+            if (labelValue.equals("null")) {
+                labelValues.add(mapKey);
+            } else {
+                labelValues.add(labelValue);
+            }
+        }
+        return labelValues.toArray(new String[0]);
+    }
+
     Gauge.Child getChild(String mapKey) {
-        return children.computeIfAbsent(
-                mapKey,
-                label -> {
-                    final ArrayList<String> labelValues = new ArrayList<>(this.labels.size());
-                    for (String labelValue : labels.values()) {
-                        if (labelValue.equals("null")) {
-                            labelValues.add(mapKey);
-                        } else {
-                            labelValues.add(labelValue);
-                        }
-                    }
-                    return collector.labels(labelValues.toArray(new String[0]));
-                });
+        return children.computeIfAbsent(mapKey, label -> collector.labels(getLabelValues(mapKey)));
     }
 
     @Override
     public void update(Map<String, Double> value) {
+        // TODO: save current labels to state backend and checkpoint to make sure
+        //  invalid gauge children can be removed after fail-over.
+        Set<String> labelsToRemove = new HashSet<>();
```

Comment:
It may be simpler to do the following.
```
final HashSet<String> labelsToRemove = new HashSet<>(children.keySet());
labelToRemove.removeAll(value.keySet());
```